core/shared/src/main/scala/org/typelevel/idna4s/core/uts46/CodePointMapper.scala

```diff
@@ -0,0 +1,449 @@
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package org.typelevel.idna4s.core.uts46
+
+import cats._
+import cats.collections.BitSet
+import cats.data._
+import cats.syntax.all._
+import java.lang.StringBuilder
+import org.typelevel.idna4s.core._
+import scala.annotation.tailrec
+import scala.collection.immutable.IntMap
+import scala.util.control.NoStackTrace
+
+object CodePointMapper extends GeneratedCodePointMapper {
+
+  /**
+   * Map the code points in the given `String` according to the UTS-46 mapping tables as
+   * described in section 5 of UTS-46. useStd3ASCIIRules is true and transitionalProcessing is
+   * false.
+   */
+  def mapCodePoints(input: String): Either[MappingException, String] =
+    mapCodePoints(true, false)(input)
+
+  /**
+   * As `mapCodePoints`, but throws on failure.
+   */
+  def unsafeMapCodePoints(input: String): String =
+    mapCodePoints(input).fold(
+      e => throw e,
+      identity
+    )
+
+  /**
+   * Map the code points in the given `String` according to the UTS-46 mapping tables as
+   * described in section 5 of UTS-46. This is step 1 of processing an input `String` according
+   * to UTS-46 for IDNA compatibility.
+   *
+   * @param useStd3ASCIIRules
+   *   Whether or not to use STD3 ASCII rules. UTS-46 strongly recommends this be `true`.
+   * @param transitionalProcessing
+   *   Determines if the deviation characters should be mapped or considered valid. From UTS-46,
+   *   "Transitional Processing should only be used immediately before a DNS lookup in the
+   *   circumstances where the registry does not guarantee a strategy of bundling or blocking.
+   *   Nontransitional Processing, which is fully compatible with IDNA2008, should be used in
+   *   all other cases.
+   *
+   * @see
+   *   [[https://www.unicode.org/reports/tr46/#IDNA_Mapping_Table UTS-46 Section 5]]
+   */
+  def mapCodePoints(useStd3ASCIIRules: Boolean, transitionalProcessing: Boolean)(
+      input: String): Either[MappingException, String] = {
+    val len: Int = input.length
+
+    @tailrec
+    def loop(
+        acc: StringBuilder,
+        errors: List[CodePointMappingException],
+        index: Int): Either[MappingException, String] =
+      if (index >= len) {
+        NonEmptyList
+          .fromList(errors)
+          .fold(
+            Right(acc.toString): Either[MappingException, String]
+          )(errors => Left(MappingException(errors, acc.toString)))
+      } else {
+        val value: Int = input.codePointAt(index)
+        val indexIncrement: Int = if (value >= 0x10000) 2 else 1
+
+        // We check valid then mapped first. This is for two reasons. First,
+        // we want to prioritize fast execution on the successful code
+        // path. Second, other than disallowed, valid/mapped make up the vast
+        // majority of the domain.
+
+        if (validAlways(value) || validNV8(value) || validXV8(value)) {
```

Comment:
Considering the non-trivial cost of initializing the data structures, could we add one more short-circuit fast-path that just checks if the input is ascii? The goal being a user plugging nothing but ascii into here will never have to pay the price of initialization.

Reply:
Where by "ascii", I mean the valid sub-range of ascii or whatever 😛

Reply:
Yeah, that seems reasonable. I'll add that in.